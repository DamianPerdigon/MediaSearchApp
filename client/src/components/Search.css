/* CSS custom properties (variables) defined for consistent theming throughout the application. */
:root {
    --primary-color: #f8f8f7; /* Light grey background color */
    --secondary-color: #fa5650; /* Vibrant red accent color */
    --text-color: #333; /* Dark grey text color */
    --fav-color: #ff9500; /* Orange color for favorite items */
    --border-color: #d7d7d7; /* Light grey border color */
}

/* Global body styles, setting the font and background color from the variables. */
body {
    font-family: 'Helvetica Neue', Arial, sans-serif; /* Sets a modern sans-serif font. */
    color: var(--text-color); /* Applies the text color variable. */
    background-color: var(--primary-color); /* Sets the primary background color. */
}

/* Styles for div elements, commonly used as containers. */
div {
    max-width: 800px; /* Maximum width to prevent the container from looking too wide on larger screens. */
    margin: 40px auto; /* Centers the div with automatic horizontal margins and adds vertical spacing. */
    padding: 20px; /* Padding inside the div for spacing around the content. */
    background: white; /* Solid white background color. */
    box-shadow: 0 2px 5px rgba(0,0,0,0.1); /* Adds subtle shadow for a layered effect. */
}

/* Paragraph styles, used for text content. */
p {
    color: var(--secondary-color); /* Applies the secondary color for text to stand out. */
    text-align: center; /* Centers the text within the paragraph. */
}

/* Unified styles for input fields, select dropdowns, and buttons. */
input[type="text"], select, button {
    padding: 10px; /* Padding for comfort. */
    margin: 10px 0; /* Vertical spacing between elements. */
    border: 1px solid var(--border-color); /* Border color from variable. */
    border-radius: 5px; /* Slightly rounded corners for a modern look. */
    display: block; /* Makes elements block-level for stacking vertically. */
    width: 100%; /* Full-width elements for a clean, aligned appearance. */
    box-sizing: border-box; /* Includes padding and border in the element's total width and height. */
}

/* Additional specific styles for buttons. */
button {
    padding: 15px 30px; /* Larger padding for a bigger click area. */
    font-size: 16px; /* Larger font size for readability. */
    font-weight: 600; /* Bold font weight for emphasis. */
    text-align: center; /* Center text inside the button. */
    vertical-align: middle; /* Align inline elements or table-cell box vertically. */
    line-height: 1.5; /* Spacing between lines of text. */
    border: none; /* Removes the border for button elements. */
    border-radius: 8px; /* More pronounced rounded corners for buttons. */
    transition: background-color 0.3s ease, transform 0.2s ease; /* Smooth transition for hover effects. */
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1); /* Adds shadow to buttons for depth. */
    cursor: pointer; /* Changes cursor to indicate clickable items. */
    margin: 10px 0; /* Vertical margin for spacing. */
    display: inline-block; /* Allows setting width and vertical margins/padding. */
    width: auto; /* Auto width to wrap content. */
    white-space: nowrap; /* Prevents text wrapping in the button. */
}

/* Hover and focus styles for button elements to provide visual feedback. */
button:hover, button:focus {
    background-color: darken(var(--secondary-color), 10%); /* Darkens the button background color on hover/focus. */
    transform: translateY(-2px); /* Slight upward movement for a 'pop' effect. */
    outline: none; /* Removes the default focus outline. */
}

/* Specific style for 'favorite' buttons, using the fav-color variable. */
button.fav-button {
    background-color: var(--fav-color); /* Orange background for favorites. */
    color: white; /* White text for contrast. */
}

/* Specific styles for the 'search' button. */
.search-button {
    background-color: var(--secondary-color); /* Vibrant red background for search button. */
    color: white; /* White text for contrast. */
    width: 100%; /* Full width for a block-level appearance. */
}

/* Media query for responsiveness on smaller screens, such as tablets. */
@media (max-width: 768px) {
    button {
        padding: 10px 20px; /* Smaller padding for smaller screens. */
        font-size: 14px; /* Smaller font size for smaller screens. */
    }
}
