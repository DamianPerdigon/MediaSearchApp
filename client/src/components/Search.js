// Import React hooks and Search component CSS.
import React, { useState, useEffect } from 'react';
import "./Search.css";

// Search component definition.
function Search() {
    // State variables for search term, media type, items list, favorites, organized favorites, loading, and error message.
    const [searchTerm, setSearchTerm] = useState('');
    const [mediaType, setMediaType] = useState('all');
    const [items, setItems] = useState([]);
    const [favorites, setFavorites] = useState([]);
    // Organized favorites state to categorize media types.
    const [organizedFavorites, setOrganizedFavorites] = useState({
        music: [],
        movie: [],
        podcast: [],
        audiobook: []
    });
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState('');

    // Effect hook to load favorites from local storage on component mount.
    useEffect(() => {
        const savedFavorites = JSON.parse(localStorage.getItem('favorites')) || [];
        setFavorites(savedFavorites);
    }, []);

    // Effect hook to save favorites to local storage when favorites state changes.
    useEffect(() => {
        localStorage.setItem('favorites', JSON.stringify(favorites));
    }, [favorites]);

    // Effect hook to organize favorites into categories based on media type.
    useEffect(() => {
        setOrganizedFavorites(organizeFavorites(favorites));
    }, [favorites]);

    // Function to categorize favorites into their respective media types.
    function organizeFavorites(favs) {
        return favs.reduce((acc, favorite) => {
            const { kind, wrapperType } = favorite;
            // Categorization logic based on media type.
            if (kind === 'song' || kind === 'music-track') {
                acc.music.push(favorite);
            } else if (kind === 'feature-movie') {
                acc.movie.push(favorite);
            } else if (kind === 'podcast') {
                acc.podcast.push(favorite);
            } else if (wrapperType === 'audiobook') {
                acc.audiobook.push(favorite);
            }
            return acc;
        }, { music: [], movie: [], podcast: [], audiobook: [] });
    }

    // Async function to fetch items based on search term and media type.
    async function fetchItems() {
        // Prevent fetch on empty search term.
        if (!searchTerm.trim()) return;

        // Set loading state and clear error state.
        setLoading(true);
        setError('');

        try {
            // Perform fetch request and process response.
            const response = await fetch(`/search?term=${encodeURIComponent(searchTerm)}&media=${mediaType}`);
            if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
            const data = await response.json();
            setItems(data.results || []);
        } catch (e) {
            // Set error state and log the error.
            setError('There was a problem fetching data.');
            console.error(e);
        } finally {
            // Reset loading state.
            setLoading(false);
        }
    }

    // Function to add an item to the favorites list.
    function addToFavorites(item) {
        setFavorites(favs => (!favs.some(f => f.trackId === item.trackId) ? [...favs, item] : favs));
    }

    // Function to remove an item from the favorites list.
    function removeFromFavorites(trackId) {
        setFavorites(favs => favs.filter(f => f.trackId !== trackId));
    }

    // Handler for search term changes with a debounce effect.
    function handleSearchTermChange(e) {
        setSearchTerm(e.target.value);
        debounce(fetchItems, 500)();
    }

    // Render the Search component UI.
    return (
        <div>
            {error && <p>Error: {error}</p>}
            {loading && <p>Loading...</p>}
            {/* Search input and media type selector */}
            <input
                type="text"
                value={searchTerm}
                onChange={handleSearchTermChange}
                placeholder="Search for content"
            />
            <select value={mediaType} onChange={e => setMediaType(e.target.value)}>
                <option value="all">All</option>
                <option value="music">Music</option>
                <option value="movie">Movies</option>
                <option value="podcast">Podcasts</option>
                <option value="audiobook">Audiobooks</option>
            </select>
            <button onClick={fetchItems}>Search</button>
            {/* List of items resulting from the search */}
            <ul>
                {items.map(item => (
                    <li key={item.trackId}>
                        {item.trackName} - {item.artistName}
                        <button onClick={() => addToFavorites(item)}>Add to Favorites</button>
                    </li>
                ))}
            </ul>
            {/* Display of organized favorites */}
            <h2>Favorites</h2>
            {Object.entries(organizedFavorites).map(([key, value]) => (
                <div key={key}>
                    <h3>{key.charAt(0).toUpperCase() + key.slice(1)}</h3>
                    <ul>
                        {value.map(favorite => (
                            <li key={favorite.trackId}>
                                {favorite.trackName || favorite.collectionName} - {favorite.artistName}
                                <button onClick={() => removeFromFavorites(favorite.trackId)}>Remove</button>
                            </li>
                        ))}
                    </ul>
                </div>
            ))}
        </div>
    );
}

// Export the Search component as the default export of this module.
export default Search;

// debounce function to limit the frequency of function execution.
function debounce(func, wait) {
    let timeout;
    return function executedFunction(...args) {
        const later = () => {
            clearTimeout(timeout);
            func(...args);
        };
        clearTimeout(timeout);
        timeout = setTimeout(later, wait);
    };
}
