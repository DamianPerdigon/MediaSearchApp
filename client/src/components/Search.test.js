// Import necessary React module and testing utility.
import React from 'react';
import { render } from '@testing-library/react';
import Search from './Search'; // Import the Search component to be tested.

// Define a mock localStorage implementation.
const localStorageMock = (function () {
    let store = {}; // Create a simple store to keep key-value pairs in memory.
    return {
        getItem: function (key) { // Mock for the getItem method of localStorage.
            return store[key] || null;
        },
        setItem: function (key, value) { // Mock for the setItem method of localStorage.
            store[key] = value.toString(); // Ensure that all values are stored as strings.
        },
        clear: function () { // Mock for the clear method of localStorage.
            store = {}; // Reset the store to an empty object.
        }
    };
})();

// Define the mock localStorage on the window object.
Object.defineProperty(window, 'localStorage', {
    value: localStorageMock // Assign the mock localStorage as the implementation on window.
});

// Mock the global fetch API.
global.fetch = jest.fn(() =>
    Promise.resolve({ // Return a resolved promise to simulate a successful fetch request.
        ok: true, // Indicate that the fetch request was successful.
        json: () => Promise.resolve({ // Mock the json method of the response.
            results: [] // Return an empty array to simulate an empty fetch response.
        }),
    })
);

// Describe block defines a test suite for the Search component.
describe('Search component', () => {
    it('renders correctly and matches snapshot', () => {
        // Render the Search component and take a snapshot.
        const { asFragment } = render(<Search />);
        // Assert that the rendered output matches the saved snapshot.
        expect(asFragment()).toMatchSnapshot();
    });


});
