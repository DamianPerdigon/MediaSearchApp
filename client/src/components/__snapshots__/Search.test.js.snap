// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <input
      placeholder="Search for content"
      type="text"
      value=""
    />
    <select>
      <option
        value="all"
      >
        All
      </option>
      <option
        value="music"
      >
        Music
      </option>
      <option
        value="movie"
      >
        Movies
      </option>
      <option
        value="podcast"
      >
        Podcasts
      </option>
      <option
        value="audiobook"
      >
        Audiobooks
      </option>
    </select>
    <button>
      Search
    </button>
    <ul />
    <h2>
      Favorites
    </h2>
    <div>
      <h3>
        Music
      </h3>
      <ul />
    </div>
    <div>
      <h3>
        Movie
      </h3>
      <ul />
    </div>
    <div>
      <h3>
        Podcast
      </h3>
      <ul />
    </div>
    <div>
      <h3>
        Audiobook
      </h3>
      <ul />
    </div>
  </div>
</DocumentFragment>
`;
